purchases-sample-java/src/main/java/com/revenuecat/purchases/purchases_sample_java/UpsellActivity.java

```diff
@@ -0,0 +1,176 @@
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_upsell);
+
+        monthlyPurchaseView = findViewById(R.id.monthly_purchase);
+        annualPurchaseView = findViewById(R.id.annual_purchase);
+        unlimitedPurchaseView = findViewById(R.id.unlimited_purchase);
+
+        showScreen(null);
+        findViewById(R.id.skip).setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                Navigator.startCatsActivity(UpsellActivity.this, false);
+            }
+        });
+        Purchases.getSharedInstance().setUpdatedPurchaserInfoListener(new UpdatedPurchaserInfoListener() {
+            @Override
+            public void onReceived(@NonNull PurchaserInfo purchaserInfo) {
+                checkForProEntitlement(purchaserInfo);
+            }
+        });
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        Purchases.getSharedInstance().getOfferings(new ReceiveOfferingsListener() {
+            @Override
+            public void onReceived(@NonNull Offerings offerings) {
+                showScreen(offerings);
+            }
+
+            @Override
+            public void onError(@NonNull PurchasesError error) {
+                Log.e("Purchases Sample", error.getMessage());
+            }
+        });
+    }
+
+    private void showScreen(@Nullable Offerings offerings) {
+        if (offerings == null) {
+            monthlyPurchaseView.setText("Loading...");
+            monthlyPurchaseView.setEnabled(false);
+            annualPurchaseView.setText("Loading...");
+            annualPurchaseView.setEnabled(false);
+        } else {
+            Offering currentOffering = offerings.getCurrent();
+            if (currentOffering != null) {
+                setupMonthlyButton(currentOffering.getMonthly());
+                setupAnnualButton(currentOffering.getAnnual());
+                setupUnlimitedButton(currentOffering.getLifetime());
+            } else {
+                Log.e("Purchases Sample", "Error loading current offering");
+            }
+        }
+    }
+
+    private void setupUnlimitedButton(@Nullable final Package lifetimePackage) {
+        if (lifetimePackage != null) {
+            SkuDetails unlimitedProduct = lifetimePackage.getProduct();
+            String loadedText = "Buy Unlimited - " + unlimitedProduct.getPriceCurrencyCode() + " " + unlimitedProduct.getPrice();
+            unlimitedPurchaseView.setTag(loadedText);
+            showLoading(unlimitedPurchaseView, false);
+            unlimitedPurchaseView.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    makePurchase(lifetimePackage, unlimitedPurchaseView);
+                }
+            });
+        } else {
+            Log.e("Purchases Sample", "Error loading lifetime package");
+        }
+    }
```

Comment:
there's a lot of duplicated code in the setupXButton methods, maybe it could be extracted into something like `setupButton(package, durationString, view)`?

Reply:
let me see what I can do!